src/test/java/htsjdk/samtools/cram/encoding/external/ExternalCodecEquivalenceTest.java

```diff
@@ -0,0 +1,176 @@
+public class ExternalCodecEquivalenceTest extends HtsjdkTest {
+    // show that External codecs are equivalent within certain ranges
+
+    // if all values are 0 <= v <= 0x7F (high bit 7)
+    // ExternalByteCodec, ExternalIntegerCodec, and ExternalLongCodec streams are identical
+    // because raw bytes, ITF8 and LTF8 are identical
+
+    @Test(dataProvider = "testPositiveByteLists", dataProviderClass = IOTestCases.class)
+    public void byteEquivalenceTest(final List<Byte> values) {
+        codecPairTest(values, this::externalByteCodecWrite,      this::externalIntegerCodecRead);
+        codecPairTest(values, this::externalByteCodecWrite,      this::externalLongCodecRead);
+        codecPairTest(values, this::externalIntegerCodecWrite,  this::externalByteCodecRead);
+        codecPairTest(values, this::externalIntegerCodecWrite,  this::externalLongCodecRead);
+        codecPairTest(values, this::externalLongCodecWrite,     this::externalByteCodecRead);
+        codecPairTest(values, this::externalLongCodecWrite,     this::externalIntegerCodecRead);
+    }
+
+    // if all values are 0 <= v <= 0x0F FF FF FF (high bit 28)
+    // ExternalIntegerCodec, and ExternalLongCodec streams are identical
+    // because ITF8 and LTF8 are identical
+
+    @Test(dataProvider = "testUint28Lists", dataProviderClass = IOTestCases.class)
+    public void intEquivalenceTest(final List<Integer> values) {
+        codecPairTest(values, this::externalIntegerCodecWrite,  this::externalLongCodecRead);
+        codecPairTest(values, this::externalLongCodecWrite,     this::externalIntegerCodecRead);
+    }
+
+    private <T extends Number> void codecPairTest(final List<T> values, Writer<T> writer, Reader reader) {
+        byte[] written = writer.write(values);
+        final List<Long> read = reader.read(written, values.size());
+
+        final List<Long> expected = values
+                .stream()
+                .map(Number::longValue)
+                .collect(Collectors.toList());
+
+        Assert.assertEquals(read, expected);
+    }
+
+    private interface Writer<T> {
+        byte[] write(final List<T> toWrite);
+    }
+
+    // byte, integer, and long can all be read as Long
+    private interface Reader {
+        List<Long> read(final byte[] toRead, final int count);
+    }
+
+    private <T extends Number> byte[] externalByteCodecWrite(List<T> values) {
+        byte[] written;
+        try (final ByteArrayOutputStream os = new ByteArrayOutputStream()) {
+            final CRAMCodec<Byte> writeCodec = new ExternalByteCodec(null, os);
+
+            for (final T value : values) {
+                writeCodec.write(value.byteValue());
+            }
+            os.flush();
+            written = os.toByteArray();
+        } catch (final IOException e) {
+            throw new RuntimeIOException(e);
+        }
+        return written;
+    }
+
+    private <T extends Number> byte[] externalIntegerCodecWrite(List<T> values) {
+        byte[] written;
```

Comment:
These look refactorable to reduce the size of the repeated code, but it's always hard to do that with numeric types than I think.

Reply:
Yeah, I tried to do so, but could never get the casting/conversion right